src/pytorch_kinematics/ik.py

```diff
@@ -0,0 +1,376 @@
+        self.early_stopping_no_improvement_patience = early_stopping_no_improvement_patience
+
+        self.max_iterations = max_iterations
+        self.lr = lr
+        self.regularlization = regularlization
+        self.optimizer_method = optimizer_method
+        self.line_search = line_search
+
+        self.err = None
+        self.err_all = None
+        self.err_min = None
+        self.no_improve_counter = None
+
+        self.pos_tolerance = pos_tolerance
+        self.rot_tolerance = rot_tolerance
+        self.initial_config = retry_configs
+        if retry_configs is None and num_retries is None:
+            raise ValueError("either initial_configs or num_retries must be specified")
+
+        # sample initial configs instead
+        self.config_sampling_method = config_sampling_method
+        self.joint_limits = joint_limits
+        if retry_configs is None:
+            self.initial_config = self.sample_configs(num_retries)
+        else:
+            if retry_configs.shape[1] != self.dof:
+                raise ValueError("initial_configs must have shape (N, %d)" % self.dof)
+        # could give a batch of initial configs
+        self.num_retries = self.initial_config.shape[-2]
+
+    def sample_configs(self, num_configs: int) -> torch.Tensor:
+        if self.config_sampling_method == "uniform":
+            # bound by joint_limits
+            if self.joint_limits is None:
+                raise ValueError("joint_limits must be specified if config_sampling_method is uniform")
+            return torch.rand(num_configs, self.dof, device=self.device) * (
+                    self.joint_limits[:, 1] - self.joint_limits[:, 0]) + self.joint_limits[:, 0]
+        elif self.config_sampling_method == "gaussian":
+            return torch.randn(num_configs, self.dof, device=self.device)
+        elif callable(self.config_sampling_method):
+            return self.config_sampling_method(num_configs)
+        else:
+            raise ValueError("invalid config_sampling_method %s" % self.config_sampling_method)
+
+    def solve(self, target_poses: Transform3d) -> IKSolution:
+        """
+        Solve IK for the given target poses in robot frame
+        :param target_poses: (N, 4, 4) tensor, goal pose in robot frame
+        :return: IKSolution solutions
+        """
+        raise NotImplementedError()
+
+
+def delta_pose(m: torch.tensor, target_pos, target_rot_rpy):
+    """
+    Determine the error in position and rotation between the given poses and the target poses
+
+    :param m: (N x M x 4 x 4) tensor of homogenous transforms
+    :param target_pos:
+    :param target_rot_rpy:
+    :return: (N*M, 6, 1) tensor of delta pose (dx, dy, dz, droll, dpitch, dyaw)
+    """
+    pos_diff = target_pos.unsqueeze(1) - m[:, :, :3, 3]
+    pos_diff = pos_diff.view(-1, 3, 1)
+    rot_diff = target_rot_rpy.unsqueeze(1) - rotation_conversions.matrix_to_euler_angles(m[:, :, :3, :3],
```

Comment:
Are we sure this is a good rotation distance? Just substracting euler angles isn't generally a good idea -- you have potential wrap-around/discontinuity issues

Reply:
Thanks for finding this! I replaced it with unit quaternion error (the quaternion required to get from current to desired orientation), converted to an axis angle representation. 

Note that this is for computing error in end effector position and orientation to be multiplied by the jacobian pseudoinverse, so it needs to be 6D. [\dot{x} \omega] = J \dot{q} where we are approximating it to [\Delta{x} \Delta{r}] = J \Delta{q}

It's getting much better convergence now.